mapiproxy/servers/default/emsmdb/oxcfxics.c

```diff
@@ -3480,27 +3545,58 @@ static enum MAPISTATUS oxcfxics_fill_transfer_state_arrays(TALLOC_CTX *mem_ctx,
 						  emsmdbp_get_contextID(table_object), table_object->backend_object,
 						  synccontext->properties.cValues, synccontext->properties.aulPropTag);
 		OPENCHANGE_RETVAL_IF(ret != MAPISTORE_SUCCESS, mapistore_error_to_mapi(ret), local_mem_ctx);
+		if (synccontext->request.is_collector) {
+			/* Set restriction on which change numbers and
+			   fmid we are interested in */
+			ret = oxcfxics_transfer_state_set_table_restriction(emsmdbp_ctx,
+									    table_object,
+									    synccontext);
+			OPENCHANGE_RETVAL_IF(ret != MAPISTORE_SUCCESS, mapistore_error_to_mapi(ret), local_mem_ctx);
+
+			ret = mapistore_table_get_row_count(emsmdbp_ctx->mstore_ctx, emsmdbp_get_contextID(table_object),
+							    table_object->backend_object, MAPISTORE_PREFILTERED_QUERY,
+							    &table_object->object.table->denominator);
+			OPENCHANGE_RETVAL_IF(ret != MAPISTORE_SUCCESS, mapistore_error_to_mapi(ret), local_mem_ctx);
+
+			OC_DEBUG(5, "Number of updated messages to return: %d with CN >= 0x%.16"PRIx64,
+                                 table_object->object.table->denominator, synccontext->next_cn);
+
+			OPENCHANGE_RETVAL_IF(table_object->object.table->denominator == 0,
+					     MAPI_E_SUCCESS, local_mem_ctx);
+		}
 	}
+
+	if (synccontext->request.is_collector) {
+		involved_fmids = RAWIDSET_convert_to_idset(local_mem_ctx, synccontext->involved_fmids);
+		IDSET_dump(involved_fmids, "involved fmids");
+	}
+
 	table = table_object->object.table;
 	for (i = 0; i < table->denominator; i++) {
 		data_pointers = emsmdbp_object_table_get_row_props(NULL, emsmdbp_ctx, table_object, i, MAPISTORE_PREFILTERED_QUERY, &retvals);
 		if (data_pointers) {
 			eid = *(uint64_t *) data_pointers[0];
-			emsmdbp_replid_to_guid(emsmdbp_ctx, owner, eid & 0xffff, &replica_guid);
-			RAWIDSET_push_guid_glob(sync_data->eid_set, &replica_guid, (eid >> 16) & 0x0000ffffffffffff);
-			
+			/* Check if the given eid is in the involved ones */
+			if (synccontext->request.is_collector && !IDSET_includes_guid_glob(involved_fmids, &sync_data->replica_guid, (eid >> 16) & 0x0000ffffffffffff)) {
+				OC_DEBUG(5, "Skipping putting eid (0x%.16"PRIx64") as it is not included in an upload operation", eid);
+				continue;
+			}
 			if (retvals[sync_data->prop_index.change_number]) {
 				OC_DEBUG(1, "mandatory property PidTagChangeNumber not returned for message %.16"PRIx64, eid);
 				talloc_free(local_mem_ctx);
 				return MAPI_E_CALL_FAILED;
 			}
+
+			emsmdbp_replid_to_guid(emsmdbp_ctx, owner, eid & 0xffff, &replica_guid);
```

Comment:
Check retval, it can return MAPI_E_NOT_FOUND and replica_guid to be junk because not memset/initialized beforehand. 

Additionally a memset on replica_guid to initialize its content to 0 before hand wouldn't be harmful.